pytensor/xtensor/spaces.py

```diff
@@ -0,0 +1,177 @@
+import sys
+from collections.abc import Iterable, Iterator, Sequence
+from typing import (
+    FrozenSet,
+    Protocol,
+    SupportsIndex,
+    Union,
+    cast,
+    overload,
+    runtime_checkable,
+)
+
+
+@runtime_checkable
+class DimLike(Protocol):
+    """Most basic signature of a dimension."""
+
+    def __str__(self) -> str:
+        ...
+
+    def __hash__(self) -> int:
+        ...
+
+
+class Dim(DimLike):
```

Comment:
As far as I remember, it is a bad practice to inherit a base class from the protocol

Reply:
Dim can be just a dataclass instead

Reply:
According to the explanation in the PEP I would disagree: https://peps.python.org/pep-0544/#explicitly-declaring-implementation

By inheriting the protocol, we enable type checkers to warn about incomplete/incorrect implementations.